src/main/groovy/org/transmartproject/batch/tag/TagTypeResource.groovy

```diff
@@ -0,0 +1,73 @@
+package org.transmartproject.batch.tag
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import org.springframework.batch.core.configuration.annotation.JobScope
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.jdbc.core.JdbcTemplate
+import org.springframework.jdbc.core.RowMapper
+import org.springframework.stereotype.Component
+import org.transmartproject.batch.clinical.db.objects.Tables
+
+import javax.annotation.PostConstruct
+import java.sql.ResultSet
+import java.sql.SQLException
+
+/**
+ * Structure to hold all the tag types and associated options.
+ */
+@Component
+@JobScope
+@Slf4j
+@CompileStatic
+class TagTypeResource {
+
+    @Autowired
+    JdbcTemplate jdbcTemplate
+
+    class TagTypeToOptionsMapper implements RowMapper<String> {
+
+        Map<String, Map<String, Integer>> tagTypeToOptionsMap = [:]
+
+        @Override
+        String mapRow(ResultSet rs, int rowNum) throws SQLException {
+            String tagType = rs.getString('tag_type')
+            String value = rs.getString('value')
+            Integer optionId = rs.getInt('tag_option_id')
+            def options = tagTypeToOptionsMap[tagType]
+            if (options == null) {
+                tagTypeToOptionsMap[tagType] = [:] as Map<String, Integer>
+                options = tagTypeToOptionsMap[tagType]
+            }
+            options.put(value, optionId)
+            log.debug("Add value '${value}' (${optionId}) for tag type ${tagType}")
+            value
+        }
+
+    }
+
+    private final Map<String, Map<String, Integer>> tagTypeToOptionsMap = [:]
+
+    @PostConstruct
```

Comment:
This should be done on a restart-if-complete step, not on the application context initialization (!).

Also, as I mentioned before, I think the database access methods that are now in `TagTypeResource` and `TagTypeService` should be put in one method (you can of course still have a bean where the list of tags options is stored)